### PR TITLE
prometheus: log virtual memory limits

### PR DESCRIPTION
**Pull request by simonpasquier**
See https://github.com/prometheus/prometheus/issues/4392#issuecomment-407549933

```
       RLIMIT_AS
              This is the maximum size of the process's virtual memory
              (address space).  The limit is specified in bytes, and is
              rounded down to the system page size.  This limit affects
              calls to brk(2), mmap(2), and mremap(2), which fail with the
              error ENOMEM upon exceeding this limit.  In addition, auto‐
              matic stack expansion fails (and generates a SIGSEGV that
              kills the process if no alternate stack has been made avail‐
              able via sigaltstack(2)).  Since the value is a long, on
              machines with a 32-bit long either this limit is at most
              2 GiB, or this resource is unlimited.
```

If the process has limits on the virtual memory, it may fail with `mmap: cannot allocate memory` errors. Similarly to what we have for file descriptors, printing the virtual limits at startup will help troubleshooting.

### CODE REVIEW

File: cmd/prometheus/limits_default.go [hunk 1]
```diff
@@ -21,12 +21,30 @@ import (
...
+}
+
+// FdLimits returns the soft and hard limits for file descriptors.
+func FdLimits() string {
+	return getLimits(syscall.RLIMIT_NOFILE)
+}
+
+// VmLimits returns the soft and hard limits for virtual memory.
+func VmLimits() string {
+	return getLimits(syscall.RLIMIT_AS)
```
[the remark this answers is not in the document]
Comment (simonpasquier) [reply cmd/prometheus/limits_default.go]: Bytes.

File: cmd/prometheus/limits_default.go [hunk 2]
```diff
@@ -21,12 +21,30 @@ import (
 	"syscall"
 )
 
-// FdLimits returns the soft and hard limits for file descriptors
-func FdLimits() string {
-	flimit := syscall.Rlimit{}
-	err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &flimit)
+var unlimited int64 = syscall.RLIM_INFINITY
```
[the remark this answers is not in the document]
Comment (simonpasquier) [reply cmd/prometheus/limits_default.go]: Agreed and I'd be happy to hear about a better way. `syscall.RLIM_INFINITY` is defined as `const RLIM_INFINITY = -0x1` and I didn't find a way to make Go happy with it.

Basically this code doesn't compile with `constant -1 overflows uint64`.

```go
var v := uint64(1)
if v == uint64(syscall.RLIM_INFINITY) {
...
}
```

**Comment by brancz** [comment]
Maybe it would additionally make sense to expose a metric on this?

File: cmd/prometheus/limits_default.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (brian-brazil) [reply cmd/prometheus/limits_default.go]: Can you add a comment to that effect?

File: cmd/prometheus/limits_default.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (brian-brazil) [reply cmd/prometheus/limits_default.go]: Can you put that in the output?

**Comment by brian-brazil** [comment]
:+1:

**Comment by simonpasquier** [comment]
> Maybe it would additionally make sense to expose a metric on this?

Makes sense. That should be tracked in the Prometheus client IIUC.

https://godoc.org/github.com/prometheus/client_golang/prometheus#NewProcessCollector

**Comment by brancz** [comment]
Absolutely right! Thanks!

**Pull request merged**